packages/web-components/fast-router/src/commands.ts

```diff
@@ -0,0 +1,198 @@
+import {
+    Mutable,
+    ElementStyles,
+    html,
+    HTMLView,
+    ViewTemplate,
+    FASTElementDefinition,
+    defaultExecutionContext,
+} from "@microsoft/fast-element";
+import { navigationParticipant } from "./participants";
+import { NavigationTransaction } from "./transaction";
+import { Router } from "./router";
+import { RouterConfiguration } from "./configuration";
+import { Transition } from "./transition";
+import {
+    Layout,
+    LayoutAndTransitionRouteDefinition,
+    HasElement,
+    HasTemplate,
+} from "./routes";
+import { Navigation } from "./navigation";
+
+export interface NavigationCommand {
+    execute(transaction: NavigationTransaction): Promise<void>;
+}
+
+export class Ignore implements NavigationCommand {
+    public async execute(transaction: NavigationTransaction): Promise<void> {}
+}
+
+export class Redirect implements NavigationCommand {
+    constructor(private redirect: string) {}
+
+    public async execute(transaction: NavigationTransaction): Promise<void> {
+        // TODO: generate new route based on redirect and route params
+        Navigation.replace(this.redirect);
+    }
+}
+
+export class Render implements NavigationCommand {
+    private _layout: Layout | null = null;
+    private _transition: Transition | null = null;
+
+    constructor(
+        private owner: RouterConfiguration,
+        private createView: () => Promise<HTMLView>
+    ) {}
+
+    public get transition(): Transition {
+        return this._transition || this.owner.defaultTransition;
+    }
+
+    public set transition(value: Transition) {
+        this._transition = value;
+    }
+
+    public get layout(): Layout {
+        return this._layout || this.owner.defaultLayout;
+    }
+
+    public set layout(value: Layout) {
+        this._layout = value;
+    }
+
+    public async execute(transaction: NavigationTransaction): Promise<void> {
+        if (!(await transaction.tryLeave())) {
+            return;
+        }
+
+        const router = transaction.owner;
+        let currentView = router.view;
+        let createView = this.createView;
+
+        const viewModel = {}; // TODO: build view model from route params
+
+        if (router.command === this) {
```

Comment:
So, if I understand correctly if the cope of the router.command is the same as the current scope (ie we are on the same route), then createView returns the currentView? Otherwise a new view will be created? What gets returned if the currentView is returned though, won't it still pass the if statement below?